File: backend/src/services/debtSimplifier.ts
export interface RawDebt {
  debtorId: string;
  debtorName: string;
  creditorId: string;
  creditorName: string;
  amount: number;
  splitTitle: string;
  splitId: string;
}

export interface SimplifiedDebt {
  debtorId: string;
  debtorName: string;
  creditorId: string;
  creditorName: string;
  amount: number;
  chain: RawDebt[];
}

const EPSILON = 0.005;

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

export function simplifyDebts(rawDebts: RawDebt[]): SimplifiedDebt[] {
  if (rawDebts.length === 0) return [];

  // Step 1: Net balance per person — positive means they are owed, negative means they owe
  const netBalance = new Map<string, number>();
  const names = new Map<string, string>();

  for (const debt of rawDebts) {
    netBalance.set(debt.creditorId, (netBalance.get(debt.creditorId) ?? 0) + debt.amount);
    netBalance.set(debt.debtorId, (netBalance.get(debt.debtorId) ?? 0) - debt.amount);
    names.set(debt.creditorId, debt.creditorName);
    names.set(debt.debtorId, debt.debtorName);
  }

  // Step 2: Greedy minimum cash flow — match biggest debtor with biggest creditor
  const balances = new Map(netBalance);
  const result: SimplifiedDebt[] = [];

  while (true) {
    let maxCreditorId: string | null = null;
    let maxCreditorBalance = EPSILON;
    let maxDebtorId: string | null = null;
    let maxDebtorBalance = -EPSILON;

    for (const [id, bal] of balances) {
      if (bal > maxCreditorBalance) {
        maxCreditorBalance = bal;
        maxCreditorId = id;
      }
      if (bal < maxDebtorBalance) {
        maxDebtorBalance = bal;
        maxDebtorId = id;
      }
    }

    if (!maxCreditorId || !maxDebtorId) break;

    const payment = round2(Math.min(maxCreditorBalance, -maxDebtorBalance));
    if (payment <= 0) break;

    // Step 3: Chain — all raw debts directly between these two people (either direction)
    const chain = rawDebts.filter(
      (d) =>
        (d.debtorId === maxDebtorId && d.creditorId === maxCreditorId) ||
        (d.debtorId === maxCreditorId && d.creditorId === maxDebtorId)
    );

    result.push({
      debtorId: maxDebtorId,
      debtorName: names.get(maxDebtorId)!,
      creditorId: maxCreditorId,
      creditorName: names.get(maxCreditorId)!,
      amount: payment,
      chain,
    });

    balances.set(maxCreditorId, round2(maxCreditorBalance - payment));
    balances.set(maxDebtorId, round2(maxDebtorBalance + payment));
  }

  return result;
}

File: frontend/lib/features/splits/services/debt_simplifier.dart
import '../models/debt_models.dart';

class DebtSimplifier {
  const DebtSimplifier._();

  static List<SimplifiedDebt> simplify(List<RawDebt> rawDebts) {
    if (rawDebts.isEmpty) return [];

    final names = <String, String>{};
    for (final d in rawDebts) {
      names[d.debtorId] = d.debtorName;
      names[d.creditorId] = d.creditorName;
    }

    final net = <String, double>{};
    for (final d in rawDebts) {
      net[d.creditorId] = (net[d.creditorId] ?? 0) + d.amount;
      net[d.debtorId] = (net[d.debtorId] ?? 0) - d.amount;
    }

    final debtorRaw = <String, List<RawDebt>>{};
    final creditorRaw = <String, List<RawDebt>>{};
    for (final d in rawDebts) {
      debtorRaw.putIfAbsent(d.debtorId, () => []).add(d);
      creditorRaw.putIfAbsent(d.creditorId, () => []).add(d);
    }

    final balances = net.entries
        .where((e) => e.value.abs() > 0.001)
        .map((e) => _Balance(id: e.key, amount: e.value))
        .toList();

    final result = <SimplifiedDebt>[];

    while (true) {
      balances.sort((a, b) => b.amount.compareTo(a.amount));
      final creditors = balances.where((b) => b.amount > 0.001).toList();
      final debtors = balances.where((b) => b.amount < -0.001).toList();
      if (creditors.isEmpty || debtors.isEmpty) break;

      final creditor = creditors.first;
      final debtor = debtors.last;

      final settle = creditor.amount < debtor.amount.abs()
          ? creditor.amount
          : debtor.amount.abs();

      final chainSet = <String>{};
      final chain = <RawDebt>[];
      for (final d in [
        ...?debtorRaw[debtor.id],
        ...?creditorRaw[creditor.id],
      ]) {
        final key = '${d.splitId}:${d.debtorId}';
        if (chainSet.add(key)) chain.add(d);
      }

      result.add(SimplifiedDebt(
        debtorId: debtor.id,
        debtorName: names[debtor.id]!,
        creditorId: creditor.id,
        creditorName: names[creditor.id]!,
        amount: _round(settle),
        chain: chain,
      ));

      creditor.amount -= settle;
      debtor.amount += settle;
      balances.removeWhere((b) => b.amount.abs() <= 0.001);
    }

    return result;
  }

  static double _round(double v) => (v * 100).round() / 100;
}

class _Balance {
  final String id;
  double amount;

  _Balance({required this.id, required this.amount});
}
